stlab/forest.hpp

```diff
@@ -0,0 +1,1240 @@
+/*
+    Copyright 2013 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+/**************************************************************************************************/
+
+#ifndef STLAB_FOREST_HPP
+#define STLAB_FOREST_HPP
+
+/**************************************************************************************************/
+
+#include <cassert>
+#include <cstddef>
+#include <functional>
+#include <iterator>
+
+#include <stlab/algorithm/reverse.hpp>
+#include <stlab/iterator/set_next.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+enum class forest_edge : bool { trailing, leading };
+
+constexpr auto forest_trailing_edge{forest_edge::trailing};
+constexpr auto forest_leading_edge{forest_edge::leading};
+
+/**************************************************************************************************/
+
+inline auto pivot(forest_edge e) { return forest_edge(static_cast<bool>(e) ^ 1); }
+
+template <typename I> // I models FullorderIterator
+inline void pivot(I& i) {
+    i.edge() = pivot(i.edge());
+}
+
+template <typename I> // I models FullorderIterator
```

Comment:
I think we should be consistant and use `<class` instead of `<typename` - in all of ASL we used `typename` (which is more accurate) but I've caved to the `class` conversion in the interest of matching the standard.

Reply:
Done